cli/flox-rust-sdk/src/providers/buildenv.rs

```diff
@@ -94,77 +107,347 @@ pub trait BuildEnv {
+    /// which has been previously evaluated successfully during locking,
+    /// and building the package with essentially `nix build <flake-url>#<attr-path>^*`.
+    /// We set `--option pure-eval true` to avoid improve reproducibility,
+    /// and allow the use of the eval-cache to avoid costly re-evaluations.
+    fn realise_flakes(&self, locked: &LockedPackageFlake) -> Result<(), BuildEnvError> {
+        // check if all store paths are valid, if so, return without eval
+        let all_valid = self.check_store_path(locked.locked_installable.outputs.values())?;
+        if all_valid {
+            return Ok(());
+        }
+
+        let mut nix_build_command = self.base_command();
+
+        // naïve url construction
+        let installable = {
+            let locked_url = &locked.locked_installable.locked_url;
+            let attr_path = &locked.locked_installable.locked_flake_attr_path;
+
+            format!("{}#{}^*", locked_url, attr_path)
+        };
+
+        println!("building flake package: {}", installable);
+
+        nix_build_command.arg("build");
+        nix_build_command.arg("--no-write-lock-file");
+        nix_build_command.arg("--no-update-lock-file");
+        nix_build_command.args(["--option", "pure-eval", "true"]);
+        nix_build_command.arg("--no-link");
+        nix_build_command.arg(&installable);
+
+        debug!(%installable, cmd=%nix_build_command.display(), "building catalog package:");
+
+        let output = nix_build_command
+            .output()
+            .map_err(BuildEnvError::CallNixBuild)?;
+
+        if !output.status.success() {
+            return Err(BuildEnvError::Realise2 {
+                install_id: locked.install_id.clone(),
+                message: String::from_utf8_lossy(&output.stderr).to_string(),
+            });
+        }
+
+        Ok(())
+    }
+
+    /// Realise a package from a store path.
+    /// [LockedPackageStorePath] is a locked package from a store path.
+    /// The package is realised by checking if the store paths are valid,
+    /// if the store path is not valid (and the store lacks the ability to reproduce it),
+    /// This function will return an error.
+    fn realise_store_path(&self, locked: &LockedPackageStorePath) -> Result<(), BuildEnvError> {
+        let valid = self.check_store_path([&locked.store_path])?;
+        if !valid {
+            return Err(BuildEnvError::Realise2 {
+                install_id: locked.install_id.clone(),
+                message: format!("'{}' is not available", locked.store_path),
+            });
+        }
+        Ok(())
+    }
+
+    /// Check if the given store paths are valid,
+    /// i.e. if the store paths exist in the store.
+    fn check_store_path(
+        &self,
+        paths: impl IntoIterator<Item = impl AsRef<OsStr>>,
+    ) -> Result<bool, BuildEnvError> {
+        let mut cmd = self.base_command();
+        cmd.arg("path-info").args(paths);
```

Comment:
at present the only difference is a delay of pinging whatever substituters you have on your system.
using this method as is would work, only that this might be followed up with an attempt to `nix copy` if it returns false, and then do the build.
With reference to our chat last week, i'd prefer manifest builds/custom packages to be **structurally** different _somewhere_ to allow them to be handled in a separate method altogether.

Reply:
Ok, makes sense.